src/test/java/io/cdap/plugin/batch/source/FTPBatchSourceTest.java

```diff
@@ -162,4 +162,31 @@ public void testSFTPPathWithSystemPropertiesAndSpecialCharactersInAuth() {
     Assert.assertEquals(5, config.getFileSystemProperties(collector).size());
     Assert.assertEquals(fileSystemProperties, config.getFileSystemProperties(collector));
   }
+
+  @Test
+  public void testInvalidServerFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration("ftp://demo:password@invalid_server:21", "");
+    config.validate(collector);
+    Assert.assertEquals(false, config.getConnection(collector));
+  }
+
+  @Test
+  public void testInvalidUsernamePasswordFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration("ftp://wronguser:wrongpassword@test.rebex.net:21", "");
+    config.validate(collector);
+    Assert.assertEquals(false, config.getConnection(collector));
+  }
+
+  @Test
+  public void testValidFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    config.configuration("ftp://demo:password@test.rebex.net:21", "");
+    config.validate(collector);
+    Assert.assertEquals(true, config.getConnection(collector));
```

Comment:
Yea this is a real server. I was not sure what test cases I should use

Reply:
Ideally unit tests shouldn't depend on connectivity to external servers (can result in flaky tests). Would prefer mocking out api calls/external connections instead. WDYT @tivv @fernst ?

Reply:
Since it's about "integration" like validation, it makes sense to use a mock ftp server, e.g https://mvnrepository.com/artifact/org.mockftpserver/MockFtpServer

Reply:
I have added the mock servers.